dbms/src/Functions/httpFunc.cpp

```diff
@@ -0,0 +1,210 @@
+    {
+        const auto const_column = checkAndGetColumnConstStringOrFixedString(column);
+        if (!const_column)
+            throw Exception(
+                "Failed to get " + format_direction + " format for function " + getName(), ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        return const_column->getValue<String>();
+    }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    bool isVariadic() const override { return true; }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t input_rows_count) override
+    {
+        // resulting column
+        const ColumnConst * result_type_column = checkAndGetColumnConstStringOrFixedString(block.getByPosition(arguments[3]).column.get());
+        if (!result_type_column)
+            throw Exception("Bad fourth argument of function " + getName(), ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        const DataTypePtr & return_type = getReturnTypeFromArgument(result_type_column);
+
+        // no need for communication on empty block
+        if (input_rows_count == 0)
+        {
+            block.getByPosition(result).column = return_type->createColumn();
+            return;
+        }
+
+        // endpoint address
+        const auto url_column = checkAndGetColumnConst<ColumnString>(block.getByPosition(arguments[0]).column.get());
+        if (!url_column)
+            throw Exception("Bad first argument of function " + getName(), ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        Poco::URI uri(url_column->getValue<String>());
+
+        // RPC communication format
+        const String rpc_input_format = getFormatFromArgument(block.getByPosition(arguments[1]).column.get(), "input");
+        const String rps_output_format = getFormatFromArgument(block.getByPosition(arguments[2]).column.get(), "output");
+
+        // sample block for results of invocation
+        // @todo for JSON formats, we need to return pre-defined name of the field here. need to select appropriate
+        Block result_block;
+        result_block.insert(ColumnWithTypeAndName(return_type, "PICK_UP_SUITABLE_NAME_HERE"));
+
+        // preparing request
+        Block argument_block;
+        for (size_t arg = 4; arg < arguments.size(); arg++)
+        {
+            argument_block.insert(block.getByPosition(arguments[arg]));
+        }
+
+        ReadWriteBufferFromHTTP::OutStreamCallback out_stream_callback = [&](std::ostream & ostr)
+        {
+            WriteBufferFromOStream out_buffer(ostr);
+            auto output = context.getOutputFormat(rpc_input_format, out_buffer, argument_block.cloneEmpty());
+            output->writePrefix();
+            output->write(argument_block);
+            output->writeSuffix();
+            output->flush();
+        };
+
+        std::unique_ptr<ReadWriteBufferFromHTTP> in_ptr;
+
+        try
+        {
+            in_ptr = std::make_unique<ReadWriteBufferFromHTTP>(
```

Comment:
It looks like it will establish new connection for each block of data multiplyed by number of working threads.
Reestablishing connections is very slow (connection pool & http keepalive is needed). If clickhouse will do it for every block times number of threads times number of simultaneous queries - for webserver it will look like a DOS attack :) and it will die fast :)

Reply:
I agree that it **can** be a possible problem. What's you consider a possible solution here? Have a connection pool for httpFunc, with setting like http_func_connection_pool_capacity?
For a single server it can be an option. For big CH cluster, I have no idea how to limit  number of connections globally (anyway, such case will require balancer between CH and application server).

Reply:
As a bonus, we will have full-featured DDoS tool inside ClickHouse :)

Reply:
:)
IMO it will just be too slow if you will reconnect to http for each block (65k rows) . It should use persistant connections pool at least for a query context.

Limiting number of connections in distributed case is another problem, which rather has no good solution.